kernel/src/transaction.rs

```diff
@@ -72,7 +96,9 @@ impl Transaction {
             engine,
             self.operation.as_deref(),
             engine_commit_info.as_ref(),
-        )?));
+        )));
+        let adds = generate_adds(engine, self.write_metadata.iter().map(|a| a.as_ref()));
+        let actions = chain(actions, adds);
```

Comment:
If you just let `actions` above be (note rename):
```rust
let commit_info = generate_commit_info(
            engine,
            self.operation.as_deref(),
            engine_commit_info.as_ref(),
        );
```

Then here you can do:
```suggestion
        let actions = chain(iter::once(commit_info), adds);
```

Reply:
yup that's cleaner!